app/src/main/java/com/hedvig/app/feature/referrals/ui/redeemcode/RedeemCodeViewModel.kt

```diff
@@ -1,28 +1,60 @@
 package com.hedvig.app.feature.referrals.ui.redeemcode
 
-import androidx.lifecycle.MutableLiveData
 import androidx.lifecycle.ViewModel
 import androidx.lifecycle.viewModelScope
 import com.hedvig.android.owldroid.graphql.RedeemReferralCodeMutation
+import com.hedvig.app.feature.offer.model.QuoteCartId
+import com.hedvig.app.feature.offer.usecase.EditCampaignUseCase
 import com.hedvig.app.feature.referrals.data.RedeemReferralCodeRepository
-import e
+import com.hedvig.app.util.featureflags.Feature
+import com.hedvig.app.util.featureflags.FeatureManager
+import kotlinx.coroutines.flow.MutableStateFlow
+import kotlinx.coroutines.flow.StateFlow
+import kotlinx.coroutines.flow.asStateFlow
+import kotlinx.coroutines.flow.update
 import kotlinx.coroutines.launch
 
 class RedeemCodeViewModel(
-    private val redeemReferralCodeRepository: RedeemReferralCodeRepository
+    private val quoteCartId: QuoteCartId?,
+    private val redeemReferralCodeRepository: RedeemReferralCodeRepository,
+    private val featureManager: FeatureManager,
+    private val editCampaignUseCase: EditCampaignUseCase,
 ) : ViewModel() {
 
-    val redeemCodeStatus: MutableLiveData<RedeemReferralCodeMutation.Data> = MutableLiveData()
+    data class ViewState(
+        val quoteCartId: QuoteCartId? = null,
+        val data: RedeemReferralCodeMutation.Data? = null,
+        val loading: Boolean = false,
+        val errorMessage: String? = null
+    )
+
+    private val _viewState = MutableStateFlow(ViewState())
+    val viewState: StateFlow<ViewState> = _viewState.asStateFlow()
 
     fun redeemReferralCode(code: String) {
         viewModelScope.launch {
-            val response =
-                runCatching { redeemReferralCodeRepository.redeemReferralCode(code) }
-            if (response.isFailure) {
-                response.exceptionOrNull()?.let { e(it) }
-                return@launch
+            val newState = if (featureManager.isFeatureEnabled(Feature.QUOTE_CART)) {
+                editQuoteCart(code)
+            } else {
+                redeemCode(code)
```

Comment:
I think these functions can just do `_viewState.update { it.copy... }` directly without reporting the new viewState back to the call site. This will also be true if we add more functions (maybe the error state clearing function mentioned above) where those would also need to handle it themselves instead of returning the new viewState somewhere. It'd be consistent across all of them that way.